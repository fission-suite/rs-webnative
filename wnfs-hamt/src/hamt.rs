use super::{KeyValueChange, Node, HAMT_VERSION};
use crate::Hasher;
use anyhow::Result;
use async_trait::async_trait;
use libipld::{serde as ipld_serde, Ipld};
use semver::Version;
use serde::{
    de::{DeserializeOwned, Error as DeError},
    ser::Error as SerError,
    Deserialize, Deserializer, Serialize, Serializer,
};
use sha3::Sha3_256;
use std::{collections::BTreeMap, hash::Hash, rc::Rc, str::FromStr};
use wnfs_common::{AsyncSerialize, BlockStore, Link};

//--------------------------------------------------------------------------------------------------
// Type Definitions
//--------------------------------------------------------------------------------------------------

/// Hash Array Mapped Trie (HAMT) is an implementation of an associative array that combines the characteristics
/// of a hash table and an array mapped trie.
///
/// This type wraps the actual implementation which can be found in the [`Node`](crate::private::Node).
///
/// # Examples
///
/// ```
/// use wnfs_hamt::Hamt;
///
/// let hamt = Hamt::<String, usize>::new();
/// println!("HAMT: {:?}", hamt);
/// ```
#[derive(Debug, Clone)]
pub struct Hamt<K, V, H = Sha3_256>
where
    H: Hasher,
{
    pub root: Rc<Node<K, V, H>>,
    pub version: Version,
}

//--------------------------------------------------------------------------------------------------
// Implementations
//--------------------------------------------------------------------------------------------------

impl<K, V, H: Hasher> Hamt<K, V, H> {
    /// Creates a new empty HAMT.
    ///
    /// # Examples
    ///
    /// ```
    /// use wnfs_hamt::Hamt;
    ///
    /// let hamt = Hamt::<String, usize>::new();
    /// println!("HAMT: {:?}", hamt);
    /// ```
    pub fn new() -> Self {
        Self {
            root: Rc::new(Node::default()),
            version: HAMT_VERSION,
        }
    }

    /// Creates a new `HAMT` with the given root node.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::rc::Rc;
    /// use wnfs_hamt::{Hamt, Node};
    ///
    /// let hamt = Hamt::<String, usize>::with_root(Rc::new(Node::default()));
    ///
    /// println!("HAMT: {:?}", hamt);
    /// ```
    pub fn with_root(root: Rc<Node<K, V, H>>) -> Self {
        Self {
            root,
            version: HAMT_VERSION,
        }
    }

    /// Gets the difference between two HAMTs at the key-value level.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::rc::Rc;
    /// use wnfs_hamt::{Hamt, Node};
    /// use wnfs_common::MemoryBlockStore;
    ///
    /// #[async_std::main]
    /// async fn main() {
    ///     let store = &mut MemoryBlockStore::default();
    ///
    ///     let main_hamt = Hamt::<String, usize>::with_root({
    ///         let mut node = Rc::new(Node::default());
    ///         node.set("foo".into(), 400, store).await.unwrap();
    ///         node.set("bar".into(), 500, store).await.unwrap();
    ///         node
    ///     });
    ///
    ///     let other_hamt = Hamt::<String, usize>::with_root({
    ///         let mut node = Rc::new(Node::default());
    ///         node.set("foo".into(), 200, store).await.unwrap();
    ///         node.set("qux".into(), 600, store).await.unwrap();
    ///         node
    ///     });
    ///
    ///     let diff = main_hamt.diff(&other_hamt, store).await.unwrap();
    ///
    ///     println!("diff: {:#?}", diff);
    /// }
    pub async fn diff(
        &self,
        other: &Self,
        store: &mut impl BlockStore,
    ) -> Result<Vec<KeyValueChange<K, V>>>
    where
        K: DeserializeOwned + Clone + Eq + Hash + AsRef<[u8]>,
        V: DeserializeOwned + Clone + Eq,
        H: Clone + 'static,
    {
        super::diff(
            Link::from(Rc::clone(&self.root)),
            Link::from(Rc::clone(&other.root)),
            store,
        )
        .await
    }

    async fn to_ipld<B: BlockStore + ?Sized>(&self, store: &mut B) -> Result<Ipld>
    where
        K: Serialize,
        V: Serialize,
    {
        Ok(Ipld::Map(BTreeMap::from([
            ("root".into(), self.root.to_ipld(store).await?),
            ("version".into(), ipld_serde::to_ipld(&self.version)?),
            ("structure".into(), ipld_serde::to_ipld("hamt")?),
        ])))
    }
}

#[async_trait(?Send)]
impl<K, V, H: Hasher> AsyncSerialize for Hamt<K, V, H>
where
    K: Serialize,
    V: Serialize,
{
    async fn async_serialize<S, B>(&self, serializer: S, store: &mut B) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        B: BlockStore + ?Sized,
    {
        self.to_ipld(store)
            .await
            .map_err(SerError::custom)?
            .serialize(serializer)
    }
}

impl<'de, K, V> Deserialize<'de> for Hamt<K, V>
where
    K: DeserializeOwned,
    V: DeserializeOwned,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Ipld::deserialize(deserializer).and_then(|ipld| ipld.try_into().map_err(DeError::custom))
    }
}

impl<K, V> TryFrom<Ipld> for Hamt<K, V>
where
    K: DeserializeOwned,
    V: DeserializeOwned,
{
    type Error = String;

    fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
        match ipld {
            Ipld::Map(mut map) => {
                let root = Rc::new(
                    Node::<K, V>::deserialize(map.remove("root").ok_or("Missing root")?)
                        .map_err(|e| e.to_string())?,
                );

                let version = match map.get("version").ok_or("Missing version")? {
                    Ipld::String(v) => Version::from_str(v).map_err(|e| e.to_string())?,
                    _ => return Err("`version` is not a string".into()),
                };

                Ok(Self { root, version })
            }
            other => Err(format!("Expected `Ipld::Map`, got {other:#?}")),
        }
    }
}

impl<K, V, H: Hasher> Default for Hamt<K, V, H> {
    fn default() -> Self {
        Self::new()
    }
}

impl<K, V, H> PartialEq for Hamt<K, V, H>
where
    K: PartialEq,
    V: PartialEq,
    H: Hasher,
{
    fn eq(&self, other: &Self) -> bool {
        self.root == other.root && self.version == other.version
    }
}

//--------------------------------------------------------------------------------------------------
// Tests
//--------------------------------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use wnfs_common::{dagcbor, MemoryBlockStore};

    #[async_std::test]
    async fn hamt_can_encode_decode_as_cbor() {
        let store = &mut MemoryBlockStore::default();
        let root = Rc::new(Node::default());
        let hamt: Hamt<String, i32> = Hamt::with_root(root);

        let encoded_hamt = dagcbor::async_encode(&hamt, store).await.unwrap();
        let decoded_hamt = dagcbor::decode::<Hamt<String, i32>>(encoded_hamt.as_ref()).unwrap();

        assert_eq!(hamt, decoded_hamt);
    }
}
