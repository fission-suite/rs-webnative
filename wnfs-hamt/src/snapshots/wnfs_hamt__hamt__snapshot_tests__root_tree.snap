---
source: wnfs-hamt/src/hamt.rs
expression: mock_ptr
---
{
  "value": {
    "root": [
      [
        255,
        255
      ],
      [
        {
          "/": "baguqeera7su5yivo2fexo46akh2l6oqjknto26bilq7h5wwghublbdoh37bq"
        },
        {
          "/": "baguqeeratakc3g2xxguocjsxkzaclgjxev6kg2xpxueflne3ak7oxwkqyv6a"
        },
        {
          "/": "baguqeeralm2kwvr2e5p5khxtntwlw2i2w5tt2ecm6hp2cz5t7dbjbzdteylq"
        },
        [
          [
            [
              22,
              0,
              0,
              0
            ],
            "22"
          ],
          [
            [
              77,
              0,
              0,
              0
            ],
            "77"
          ]
        ],
        {
          "/": "baguqeeramslree7kobdey5gskykbcyxh4rwnydk2zusp5xw5chytzx2hu3mq"
        },
        {
          "/": "baguqeeraxaeardupah3thcpycodsoafcbqorlceifbh5g7faqfy7qmqjtfpa"
        },
        {
          "/": "baguqeeram3kyaz3ftsxboqrtuck6aira5lowno7nsvrfwvzevnjragni7bga"
        },
        {
          "/": "baguqeeraivaama6yjdyircspzmj543glu7rq5q57nwxuml2gcg3ihsidez2q"
        },
        {
          "/": "baguqeera3ruklahg7rlwswiubecrwimaabjsz4erabdtsg2zagyugbaxtw2q"
        },
        {
          "/": "baguqeerao4sodh4374rc23q6s7et7inpxh36jrzt3rn3ui7trowiyrlmiujq"
        },
        {
          "/": "baguqeerakjuq7pvie62b2qawzibcoarnpm5bmswpfl2zhaxwftbsoxjysw3a"
        },
        [
          [
            [
              13,
              0,
              0,
              0
            ],
            "13"
          ],
          [
            [
              60,
              0,
              0,
              0
            ],
            "60"
          ],
          [
            [
              28,
              0,
              0,
              0
            ],
            "28"
          ]
        ],
        {
          "/": "baguqeera7pwlxurfxchcqyrnsoxpetg3um3itxsokyjuooxfteypiichst5q"
        },
        {
          "/": "baguqeeramsemmqpc2ps3zrrbquwu5n3c3sywgpaa2b6xct5yfjmvj5fandpq"
        },
        [
          [
            [
              68,
              0,
              0,
              0
            ],
            "68"
          ],
          [
            [
              48,
              0,
              0,
              0
            ],
            "48"
          ]
        ],
        {
          "/": "baguqeeraqxpcgf7sqppbf7fkwh4sykjgv33uyfwn2x4i5vo77m2i2gunsfla"
        }
      ]
    ],
    "structure": "hamt",
    "version": "0.1.0"
  },
  "cbor": "o2Ryb290goIY/xj/kNgqWCYAAakCEiD8qdwirtFJd3PAUfS/OglTZu14KFw+ftrGPQKwjcffw9gqWCYAAakCEiCYFC2bV7mo4SZXVkAlmTclfKNq770IVbSbAr7r2VDFfNgqWCYAAakCEiBbNKtWOidf1R7zbOy7aRq3Zz0QTPHfoWez+MKQ5HMmF4KChBYAAABiMjKChBhNAAAAYjc32CpYJgABqQISIGSXEhPqcEZMdNJWFBFi5+Rs3A1azST+3t0R8TzfR6bZ2CpYJgABqQISILgICI6PAfczifgThycAogwdFYiIKE/TfKCBcfgyCZle2CpYJgABqQISIGbVgGdlnK4XQjOgleAiIOrdZrvtlWJbVySrUxAZqPhM2CpYJgABqQISIEVABgPYSPCIik/LE95sy6fjDsO/ba9GL0YRtoPJAyZ12CpYJgABqQISINxopYDm/FdpWRQJBRshgABTLPCRAEc5G1kBsUMEF5212CpYJgABqQISIHck4Z+b/yItbh6XyT+hr7n35Mcz3Fu6I/OLrIxFbEUT2CpYJgABqQISIFJpD76oJ7QdQBbKAicCLXs6FkrPKvWTgvYswyddOJW2g4KEDQAAAGIxM4KEGDwAAABiNjCChBgcAAAAYjI42CpYJgABqQISIPvsu9IluI4oYi2TrvJM26M2id5OVhNHOuWZMPQgR5T72CpYJgABqQISIGSIxkHi0+W8xiGFLU63YtyxYzwA0H1xT7gqWVT0oGjfgoKEGEQAAABiNjiChBgwAAAAYjQ42CpYJgABqQISIIXeIxfyg94S/Kqx+SwpJq73TBbN1fiO1d/7NI0ajZFWZ3ZlcnNpb25lMC4xLjBpc3RydWN0dXJlZGhhbXQ="
}
